File: python/oop/abstraction.py
from abc import ABC, abstractmethod

class Account(ABC):
    def __init__(self, account_num, balance):
        self._account_num = account_num
        self._balance = balance
    
    @abstractmethod
    def withdraw(self, amount):
        pass

    @abstractmethod
    def deposit(self, amount):
        pass

    def get_account_num(self):
        return self._account_num

    def get_balance(self):
        return self._balance
    
class SavingsAccount(Account):
    def deposit(self, amount):
        self._balance += amount
        print(f"Deposited {amount} in Savings Account.")

    def withdraw(self, amount):
        if self._balance > amount : 
            self._balance -= amount
            print(f"Withdrew {amount} from Savings Account.")
        else:
            print("Insufficient Funds in Savings Account.")
    
class CurrentAccount(Account):
    def deposit(self, amount):
        self._balance += amount
        print(f"Deposited {amount} in Current Account.")

    def withdraw(self, amount):
        if self._balance > amount : 
            self._balance -= amount
            print(f"Withdrew {amount} from Current Account.")
        else:
            print("Insufficient Funds in Current Account.")

class CheckingAccount(Account):
    def deposit(self, amount):
        self._balance += amount
        print(f"Deposited {amount} in Checking Account.")

def main():
    my_savings = SavingsAccount("1234 5678 90", 500.0)
    my_savings.deposit(200.0)
    my_savings.withdraw(100.0)
    print(f"Savings Account Balance: {my_savings.get_balance()}\n")

    my_current = CurrentAccount("1234 5678 90", 1000.0)
    my_current.deposit(500.0)
    my_current.withdraw(1500.0)
    print(f"Current Account Balance: {my_current.get_balance()}\n")

    # This will give you error because withdraw method wasn't implemented for CheckingAccount.
    # my_checking = CheckingAccount("1234 5678 90", 2000)
    # my_checking.deposit(100.0)

if __name__ == "__main__":
    main()

File: cpp/oop/abstraction.cpp
#include <iostream>
#include <string>

class Account{
    protected:
        std::string account_num;
        float balance;
    public:
        Account(std::string acc_num, float bal) : account_num(acc_num), balance(bal) {}

        virtual void deposit(float amount) =0;
        virtual void withdraw(float amount) =0;

        std::string get_account_num(){
            return account_num;
        }

        float get_balance(){
            return balance; 
        }     
};

class SavingsAccount : public Account{
    public : 
        SavingsAccount(std::string acc_num, float balance) : Account(acc_num, balance) {}
    
    void deposit(float amount) override{
        balance += amount;
        std::cout<<"Deposit "<<amount<<" to Savings Account."<<std::endl;
    } 

    void withdraw(float amount) override{
        if (balance >= amount){
            balance -= amount;
            std::cout<<"Withdrew "<<amount<<" from Savings Account."<<std::endl;
        } else{
            std::cout<<"Insufficient funds in Savings Account"<<std::endl;
        }
    }
};

class CurrentAccount : public Account {
    public:
        CurrentAccount(std::string acc_num, float bal) : Account(acc_num, balance) {}
    
    void deposit(float amount){
        balance += amount;
        std::cout<<"Deposit "<<amount<<" to Current Account."<<std::endl;
    }

    void withdraw(float amount){
        if (balance >= amount){
            balance -= amount;
            std::cout<<"Withdrew "<<amount<<" from Current Account."<<std::endl;
        } else{
            std::cout<<"Insufficient funds in Current Account"<<std::endl;
        }
    }
};

int main(){
    Account* my_savings = new SavingsAccount("1234 5678 90", 100);
    my_savings->deposit(1000);
    my_savings->withdraw(500);
    std::cout<<"Savings Account Balance :"<<my_savings->get_balance()<<std::endl<<std::endl;

    Account* my_current = new CurrentAccount("1234 5678 90", 1000);
    my_current->deposit(1000);
    my_current->withdraw(2500);
    my_current->withdraw(200);
    std::cout<<"Current Account Balance :"<<my_current->get_balance()<<std::endl;

    return 0;
};
